api/src/main/java/org/apache/iceberg/VariantLike.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import org.apache.iceberg.types.Type;
+
+/** Interface for accessing Variant fields. */
+public interface VariantLike {
+  int size();
+
+  /**
+   * Retrieves the value of the current variant element based on the provided `javaClass` type. The
+   * `javaClass` parameter should be the `javaClass` field value of an Iceberg data type TypeID
+   * {@link Type#typeId()}, such as `Boolean.class`.
+   *
+   * <p>If the current variant element holds a primitive value that can be extracted or promoted to
+   * match the specified `javaClass`, the method will return that value. The method errors out if
+   * the value is an object or fails to cast to the desired type.
```

Comment:
Why would this cast the value?  I think that we need to be careful about casting values in the storage layer. We don't want to modify values at all, nor do we want to accidentally convert values before comparing them. In general, casting should be done in engines, not in Iceberg.

Reply:
The class is removed.